src/cpp/session/modules/rmarkdown/NotebookDocQueue.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+ * NotebookDocQueue.cpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "NotebookDocQueue.hpp"
+
+#include <boost/foreach.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+using namespace rstudio::core;
+
+#define kDocQueueId             "doc_id"
+#define kDocQueueJobDesc        "job_desc"
+#define kDocQueuePixelWidth     "pixel_width"
+#define kDocQueueCharWidth      "pixel_width"
+#define kDocQueueUnits          "units"
+#define kDocQueueMaxUnits       "max_units"
+#define kDocQueueCommitMode     "commit_mode"
+#define kDocQueueCompletedUnits "completed_units"
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+NotebookDocQueue::NotebookDocQueue(const std::string& docId, 
+      const std::string& jobDesc, CommitMode commitMode,
+      int pixelWidth, int charWidth) :
+      docId_(docId),
+      jobDesc_(jobDesc),
+      commitMode_(commitMode),
+      pixelWidth_(pixelWidth),
+      charWidth_(charWidth)
+{
+}
+
+NotebookDocQueue::NotebookDocQueue()
```

Comment:
Some of the members will be garbage on this initialization path (e.g. the integers). Should you have an `empty()` method to test for a valid object?